Client fault tolerant connection by multiple hosts

df530 opened the pull request:
Changelog category (leave one):

New Feature
Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Parameter `--host` can accept multiple hosts. In case of unavailability of one of them, the client will try to connect to the next one.

Detailed description / Documentation draft:
Feature accept setting multiple hosts for connection. It approves fault tolerance of connection because when one host isn't alive client can try to connect to another one.

Parameter `--host` can accept multiple addresses. Every address must contain resolvable host and can also contain portif it isn't set, port from `--port` is used). Example of usage:
`clickhouse-client --host host1:1 host2 host3:3 --port 2 --query "SELECT 1"`

When multiple addresses are passed to the host argument and the first one is unavailable, the client will try to connect to the next one (round robin approach). In the example, if client can't connect to `host1:1`, it will try to connect to `host2` and port `2`, which is set in `--port`.

> By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

> If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.


> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

df530 committed:
round robin connection on fail

df530 committed:
Fix codestyle

amosbird committed:
Disable projection when there is JOIN or SAMPLE

df530 committed:
Add tests, remove saving host in config

PolyProgrammist commented on programs/client/Client.cpp:
You set it here, but anyway you do it later in a round robin loop. Do you really need it here?

PolyProgrammist reviewed (commented):
Please add description

df530 commented on programs/client/Client.cpp:
I need to set "host" before round-robin, because I need to create `ConnectionParameters(config())` in `connect` method. I thought, that it would be better to set `config()` before calling `connect`.

df530 committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into client-fault-tolerant-connection

df530 committed:
Fix codestyle

Avogar committed:
Add schema inference for values() table function

Avogar committed:
Remove commented line

Avogar committed:
Fix style

df530 committed:
Update programs/client/Client.cpp

Fix `--host` parameter description

Co-authored-by: Kruglov Pavel <48961922+Avogar@users.noreply.github.com>

df530 committed:
Add function which gets port from configuration

df530 committed:
Merge branch 'client-fault-tolerant-connection' of github.com:ManagedDatabases/ClickHouse into client-fault-tolerant-connection

df530 committed:
Fix errors in build check

df530 committed:
Fix error with IPv6 in host param and impove HostPort reading from input stream

df530 committed:
Fix codestyle

Avogar committed:
Support UUID in MsgPack format

Avogar committed:
FIx error messages

Avogar committed:
Add table function format(format_name, data)

Commit:
Add submodule minizip-ng

Commit:
Add utility classes ZipArchiveReader and ZipArchiveWriter.

Avogar committed:
Fix test

kitaisreal committed:
Updated sort to bitsetsort

kitaisreal committed:
Fixed tests

kitaisreal committed:
Fixed tests

kitaisreal committed:
bitsetsort fix reverse_conditional_swap

azat committed:
Fix current_user/current_address for interserver mode

Before this patch current_user/current_address will be preserved from
the previous query.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

Commit:
Add helper class to cache the result of checking access.

andycol committed:
Update clickhouse-keeper.md

fix the run command and add example

Felixoid committed:
Replace custom formats by `short` alias

kitaisreal committed:
Fixed tests

kssenii committed:
add query id for delayed interactive

KochetovNicolai committed:
Update src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

KochetovNicolai committed:
Merge branch 'master' into revert-34212-revert-33291-add-pool-to-s3-write-buffer

kitaisreal committed:
Fixed tests

KochetovNicolai committed:
Probably fix something

CurtizJ committed:
support async inserts in clickhouse-client for queries with inlined data

amosbird committed:
Disable projection for high-order storages

amosbird committed:
Fix tests when projection is enabled

Avoid using count() in quota related tests

count() can subject to many optimization techniques, which is unstable
for testing quota usage.

amosbird committed:
Adapt minmax_count_projection with ModuleLegacy

amosbird committed:
Bug fix and improvement of minmax_count_projection

amosbird committed:
Enable projection by default

kitaisreal committed:
Fixed tests

FArthur-cmd committed:
correct config

codyrobert committed:
Update and add new team members (#34268)

alexey-milovidov committed:
Merge pull request #34240 from ClickHouse/list-versions

Update list-versions.sh, update version_date.tsv

Commit:
Add 'clickhouse-client --password' comment to the scripts used in Quick Start Guide https://clickhouse.com/#quick-start

kitaisreal committed:
Fixed tests

kitaisreal committed:
Merge pull request #34225 from vitlibar/fix-checking-grants-for-show-grants

Fix checking grants for SHOW GRANTS

kitaisreal committed:
Merge pull request #34167 from azat/cmake-cleanup

Cmake leftovers cleanup

alexey-milovidov committed:
Merge pull request #34270 from mfursov/clickhouse-client-password

Add 'clickhouse-client --password' comment to the scripts used in Qui…

kitaisreal committed:
Merge pull request #34243 from azat/use_skip_indexes_if_final

Disable data skipping indexes by default for queries with FINAL

alexey-milovidov committed:
Update CertificateReloader.cpp

alexey-milovidov committed:
Add a test for shebang

alexey-milovidov committed:
Fix progress bar width

alexey-milovidov committed:
Update ClientBase.cpp

alexey-milovidov committed:
Merge pull request #34267 from CurtizJ/async-insert-in-client

Support async inserts in clickhouse-client for queries with inlined data

alexey-milovidov committed:
Fix test

alexey-milovidov committed:
Add a test

alexey-milovidov committed:
Merge pull request #34263 from azat/interserver-mode-fix

Fix current_user/current_address for interserver mode

Commit:
Translate zh/faq/general/contribute: rename old file

Commit:
Translate zh/faq/general/contribute: reimport file

Commit:
Translate zh/faq/general/contribute: translate to zh

Commit:
Translate zh/faq/general/contribute: remove old files

KochetovNicolai committed:
Merge pull request #34221 from ClickHouse/revert-34201-revert-34189-less-logging-for-remote_fs_execute_merges_on_single_replica_time_threshold

Revert "Revert "Additionally check remote_fs_execute_merges_on_single_replica_time_threshold inside ReplicatedMergeTreeQueue""

tavplubix committed:
make restarting thread less bad

KochetovNicolai committed:
Merge pull request #34274 from ClickHouse/shebang-test

Add a test for shebang

kitaisreal committed:
Merge pull request #34275 from ClickHouse/fix-progress-bar

Fix progress bar width

mergify[bot] committed:
Merge branch 'master' into ttest_confidence_intervals

KochetovNicolai committed:
Just a tiny fix.

kitaisreal committed:
Fix fractional_progress_bar test

amosbird committed:
Disable projection when doing parallel replica reading

Felixoid committed:
Run the StyleCheck always even if tests skipped

CurtizJ committed:
minor fixes

KochetovNicolai committed:
Wait for pipeline to be destroyed in case of exception

danlark1 committed:
Optimize quantilesExact{Low,High} to use nth_element instead of sort

Felixoid committed:
Mark `Run Check` as success for some cases

Felixoid committed:
Fix style for tags_stable.yml

amosbird committed:
Fix aggregation_in_order with normal projection

Felixoid committed:
Add verbosity to stylecheck

alexey-milovidov committed:
Update SECURITY.md

alesapin committed:
Merge pull request #34283 from ClickHouse/style-check-always

Tune `Run Check` and `Style Check` jobs

FArthur-cmd committed:
Comment dhParamsFile. It should be uncommented only if it has correct path

mergify[bot] committed:
Merge branch 'master' into master

evillique committed:
Merge pull request #34277 from cnmade/PF202202031430

Translate zh/faq/general/contribute

CurtizJ committed:
Merge pull request #32748 from CurtizJ/read-in-order-fixed-prefix

Support `optimize_read_in_order` if prefix of sorting key is already sorted

kitaisreal committed:
Fixed tests

Felixoid committed:
Fix check-workflows in style check, iterate over checks

kitaisreal committed:
Merge pull request #34282 from kitaisreal/fractional-progress-bar-fix-test

Fix fractional_progress_bar test

Felixoid committed:
Check style of workflows/release.yml

kitaisreal committed:
Fixed tests

kitaisreal committed:
Merge pull request #34287 from danlark1/master

Optimize quantilesExact{Low,High} to use nth_element instead of sort

flickerbox-tom committed:
Add support policy page content. Add ol.default for li margin.

tavplubix committed:
cancel merges before acquiring lock for truncate

alexey-milovidov committed:
Fix terribly low performance of `LineAsString` format

alexey-milovidov committed:
Maybe better

alexey-milovidov committed:
Fix linkage

alexey-milovidov committed:
Merge pull request #31257 from FArthur-cmd/merge_15765

Merge #15765

alexey-milovidov committed:
Fix linkage

alexey-milovidov committed:
Fix parallel loading of data parts

alexey-milovidov committed:
Update ThreadPool.cpp

alexey-milovidov committed:
Merge pull request #34309 from flickerbox/tr-add-support-policy-page

Add support policy page content.

sdil committed:
Add INTERVAL STEP example in SELECT .. ORDER BY WITH FILL documentation

This corresponds to the merged PR https://github.com/ClickHouse/ClickHouse/pull/30927

alesapin committed:
Merge pull request #34289 from ClickHouse/style-check-verbosity

Add verbosity to stylecheck

KochetovNicolai committed:
Merge pull request #34219 from ClickHouse/revert-34212-revert-33291-add-pool-to-s3-write-buffer

Revert "Revert "Add pool to WriteBufferFromS3""

KochetovNicolai committed:
Merge pull request #33260 from achimbab/ttest_confidence_intervals

Add confidence intervals to ttests

tavplubix committed:
Merge pull request #34304 from ClickHouse/cancel_merges_before_truncate

Cancel merges before acquiring lock for truncate

kitaisreal committed:
Merge pull request #34310 from ClickHouse/fix-parallel-loading-parts

Fix parallel loading of data parts

kitaisreal committed:
Merge pull request #34306 from ClickHouse/line-as-string-low-performance

Fix terribly low performance of `LineAsString` format

kitaisreal committed:
Merge pull request #34284 from CurtizJ/async-insert-in-client

Minor fixes for #34267

KochetovNicolai committed:
Probably fix data race in WriteBufferFromS3 destructor.

cmsxbc committed:
Merge branch 'master' into addressToLineWithInlines

ucasfl committed:
use LowCardinality for _file and _path virtual columns in StorageFile

fix

tavplubix committed:
Merge pull request #34297 from ClickHouse/fix_restarting_thread

Try to fix race between pullLogsToQueue and RestartingThread

azat committed:
Decrease severity for "Reading ... ranges ..." log message to Trace

That way with send_logs_level='debug', you will not get verbose
information that you already has, since there is summary row:

    Selected ... parts by partition key, ... parts by primary key, ... marks by primary key, ... marks to read from ... ranges

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

kitaisreal committed:
Function mapPopulateSeries refactoring

kitaisreal committed:
Added tests

kitaisreal committed:
Added performance tests

KochetovNicolai committed:
Fix test with ttl.

kitaisreal committed:
Fixed tests

tavplubix committed:
Merge pull request #34319 from azat/mt-logging

Decrease severity for "Reading ... ranges ..." log message to Trace

KochetovNicolai committed:
Merge pull request #34316 from ClickHouse/probably-fix-data-race-in-WriteBufferFromS3

Probably fix data race in WriteBufferFromS3 destructor.

alexey-milovidov committed:
Merge pull request #34317 from ucasfl/lowcard

use LowCardinality for _file and _path virtual columns in StorageFile

Commit:
ReadBufferFromMemory now can seek backwards after been read up to EOF.

kitaisreal committed:
Fixed tests

kitaisreal committed:
Fixed tests

kitaisreal committed:
Fixed tests

kitaisreal committed:
Fixed tests

CurtizJ committed:
fix rare bug in reading of empty arrays

CurtizJ committed:
Update order-by.md

CurtizJ committed:
Merge pull request #34312 from sdil/patch-2

Add INTERVAL STEP example in SELECT .. ORDER BY WITH FILL documentation

ucasfl committed:
continue of #34317

fix

fix

taiyang-li committed:
change as requested

taiyang-li committed:
as alias functions

kitaisreal committed:
Merge pull request #34318 from kitaisreal/map-populate-series-refactoring

Function mapPopulateSeries refactoring

amosbird committed:
Better handle pre-inputs before client start

kitaisreal committed:
Added additional performance test

kssenii committed:
Merge pull request #34265 from kssenii/add-query-id-for-client-queries-file

print query id when using `--interactive` with `--queries-file` in client/local

kitaisreal committed:
Added implementation of bitsetsort from LLVM patch

Felixoid committed:
Fix wrong OK_SKIP labels

alexey-milovidov committed:
Merge pull request #34336 from amosbird/replclientimprove

Better handle pre-inputs before client start

Commit:
Add test for propagating OpenTelemetry context via gRPC protocol.

Felixoid committed:
Merge pull request #34340 from ClickHouse/ok-skip-labels

Fix wrong OK_SKIP labels

CurtizJ committed:
add unit test

bharatnc committed:
check and get column for s2 funcs

kitaisreal committed:
Merge pull request #34339 from kitaisreal/map-populate-series-refactoring

Function mapPopulateSeries added additional performance test

bharatnc committed:
check and get column remaining s2 funcs

bharatnc committed:
minor fixes

alexey-milovidov committed:
Fix wrong engine in SHOW CREATE DATABASE with engine Memory #34225

taiyang-li committed:
fix fast test

alexey-milovidov committed:
Add a test

taiyang-li committed:
add fasttest

taiyang-li committed:
make aliases case insensitive

alexey-milovidov committed:
Update clickhouse-server.service

alexey-milovidov committed:
Merge pull request #33996 from save-my-heart/fix/signal_pipe_buf_size

reduce signal_pipe_buf_size

bharatnc committed:
update src/Functions/geoToS2.cpp

bharatnc committed:
fix failing tests

taiyang-li committed:
add options for clickhouse-format

alexey-milovidov committed:
Small improvements

amosbird committed:
Revise and add more comments

kitaisreal committed:
Merge pull request #34351 from ClickHouse/small-improvements-2

Small improvements

kitaisreal committed:
Merge pull request #34334 from bigo-sg/alias_function

REGEXP_MATCHES and REGEXP_REPLACE function aliases for compatibility with PostgreSQL.

kitaisreal committed:
Merge pull request #34140 from bigo-sg/max_splits

Maxsplit argument for splitByChar.

kitaisreal committed:
Fixed tests

kssenii committed:
Fix

kitaisreal committed:
Merge pull request #34344 from bharatnc/ncb/s2-fixes

explicitly check and get columns for s2 funcs

amosbird committed:
Fix order by after aggregation

alexey-milovidov committed:
Update TableFunctionS3.cpp

mergify[bot] committed:
Merge branch 'master' into fix-s3-schema-inference

alexey-milovidov committed:
Merge pull request #34357 from kssenii/fix-clickhouse-local-use-db

Fix USE database for clickhouse-local

alexey-milovidov committed:
Update references

Commit:
Fix inserting to temporary tables via gRPC.

alexey-milovidov committed:
Merge pull request #34186 from Avogar/fix-s3-schema-inference

Fix schema inference for table function s3

alexey-milovidov committed:
Merge pull request #34024 from den-crane/patch-5

make systemd to use EnvironmentFile

kitaisreal committed:
Updated sort back to psqsort

alexey-milovidov committed:
Merge pull request #34333 from ucasfl/owcard

use LowCardinality type for _file and _path in some Storages, continue of #34317

netroby committed:
sync index: add sqlite

netroby committed:
Translate zh/engines/database-engines/sqlite:  rename old file

alexey-milovidov committed:
Merge pull request #34175 from kitaisreal/bitset-sort-performance-check

bitsetsort peformance check

amosbird committed:
Better projection format and test fixes

netroby committed:
Translate zh/engines/database-engines/sqlite:  reimport files

netroby committed:
Translate zh/engines/database-engines/sqlite: translate to zh

netroby committed:
Translate zh/engines/database-engines/sqlite: remove old file

netroby committed:
Translate zh/engines/database-engines/sqlite: fix engine parameters

Commit:
Merge pull request #34341 from vitlibar/add-test-for-propagating-opentelemetry-context-grpc

Add test for propagating OpenTelemetry context via gRPC protocol

taiyang-li committed:
move clickhouse-format code from settings to format.cpp

HeenaBansal2009 committed:
Fix clang-tidy issue

netroby committed:
Translate zh/engines/table-engines/hive:  update index and sync hive translate to zh

Avogar committed:
Merge pull request #34117 from Avogar/local

Small improvement in schema inference from stdin in local

evillique committed:
Merge pull request #34358 from cnmade/PF202202071000

sync index: add sqlite

mergify[bot] committed:
Merge branch 'master' into PF202202071031

Avogar committed:
Delete empty file DateOrDateTimeFunctionsConvertion.cpp

Avogar committed:
Merge pull request #34371 from ClickHouse/Avogar-patch-2

Delete empty file DateOrDateTimeFunctionsConvertion.cpp

alesapin committed:
Merge pull request #34327 from CurtizJ/fix-reading-empty-arrays

Try to fix rare bug in reading of empty arrays

alesapin committed:
Update clickhouse-keeper.md

alesapin committed:
Update KeyDescription.cpp

Avogar committed:
Fix tests naming, remove code duplication

Avogar committed:
Fix

Avogar committed:
Fix style and test

Felixoid committed:
Use UTC in docker images

Felixoid committed:
Update performance comparison OS version

Felixoid committed:
Make testflows a multiarch image

Commit:
Translate zh/engines/database-engines/sqlite: fix symbol

Commit:
Update docs/zh/engines/table-engines/integrations/hive.md

Co-authored-by: Nikolay Degterinsky <43110995+evillique@users.noreply.github.com>

Commit:
Translate zh/engines/database-engines/sqlite: true is not need

Avogar committed:
Fix tests

Avogar committed:
Fix

evillique committed:
Merge pull request #34360 from cnmade/PF202202071031

Translate zh/engines/database-engines/sqlite

Avogar committed:
Fix comments

CurtizJ committed:
fix consecutive backward seeks in seekable read buffers

evillique committed:
Merge pull request #34366 from cnmade/PF202202071430

Translate zh/engines/table-engines/hive:  update index and sync hive …

evillique committed:
Merge pull request #34365 from DevTeamBK/Fix-clang-tidy

Fix clang-tidy issue

CurtizJ committed:
Update BufferBase.h

Avogar committed:
Make better

evillique committed:
Better test

Avogar committed:
Fix special build

Avogar committed:
Fix style

CurtizJ committed:
Merge pull request #34228 from CurtizJ/fix-subcolumns-with-dots

Fix reading of subcolumns with dots in their names

Avogar committed:
Merge pull request #34017 from Avogar/values-func

Add schema inference for values() table function.

blinkov committed:
[website] replace Bastian's photo (#34384)

* Add files via upload

* Add files via upload

amosbird committed:
Fix tests

HeenaBansal2009 committed:
updated .clang-tidy as per Alexey's Suggestion

HeenaBansal2009 committed:
Merge branch 'ClickHouse:master' into Fix-clang-tidy

df530 committed:
Fix error with IPv6 in host param

alexey-milovidov committed:
Merge pull request #34057 from Algunenano/i34053

Less flaky test_inconsistent_parts_if_drop_while_replica_not_active

alexey-milovidov committed:
Merge pull request #34376 from CurtizJ/fix-reading-empty-arrays

Fix consecutive backward seeks in seekable read buffers

alexey-milovidov committed:
Merge pull request #34345 from ClickHouse/fix-wrong-engine-database-memory

Fix wrong engine in SHOW CREATE DATABASE with engine Memory #34225

alexey-milovidov committed:
Merge pull request #34125 from Avogar/table-from-literal

Add table function format(format_name, data)

kitaisreal committed:
Merge pull request #34349 from bigo-sg/format_parse_options

Add options for clickhouse-format.

CurtizJ committed:
Merge pull request #34229 from evillique/parser

Add composability to casting and index operators

cmsxbc committed:
Merge branch 'master' into addressToLineWithInlines

cmsxbc committed:
use `max` instead of `not any` to check all rows

varadarajkumar committed:
Method called on already moved

kitaisreal committed:
Merge pull request #34388 from DevTeamBK/Fix-clang-tidy

Fix clang tidy

df530 committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into client-fault-tolerant-connection

Avogar committed:
Merge pull request #34065 from Avogar/msgpack

Support UUID in MsgPack format

df530 committed:
Delete not used include

Avogar committed:
Fix segfault in schema inference from url

alesapin committed:
Merge pull request #34364 from vitlibar/fix-inserting-into-temp-tables-via-grpc

Fix inserting to temporary tables via gRPC.

alesapin committed:
Merge pull request #34264 from andycol/patch-1

Update clickhouse-keeper.md

alesapin committed:
Merge pull request #33933 from sunny19930321/fix-substr-zk-metadata

Better local metadata comparison with ZooKeeper metadata

Avogar committed:
Merge pull request #34405 from Avogar/fix-segfault-in-url

Fix segfault in schema inference from url

CurtizJ committed:
Merge pull request #33467 from cmsxbc/addressToLineWithInlines

add function addressToLineWithInlines

KochetovNicolai committed:
Update Settings.h

KochetovNicolai committed:
Merge pull request #34305 from amosbird/projection-fix27

Fix various issues when projection is enabled by default

kitaisreal committed:
Merge pull request #34398 from DevTeamBK/input_format

Method called on already moved

Felixoid committed:
Merge pull request #34373 from ClickHouse/docker-tz

Use UTC in docker images

Felixoid committed:
Revert "Merge pull request #34373 from ClickHouse/docker-tz"

This reverts commit efd8044ab7cd35feb4e0b1bccc63f46d42e34c82, reversing
changes made to 4bb69bcb15374b696080c5fd2fe1090dc0bec2a2.

Felixoid committed:
Merge pull request #34413 from ClickHouse/docker-tz

Revert "Merge pull request #34373 from ClickHouse/docker-tz"

Commit:
Merge pull request #34226 from vitlibar/add-submodule-minizip

Add submodule minizip

df530 committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into client-fault-tolerant-connection
